### PR TITLE
copr: implement strcmp considering collation

### LINKED ISSUE #8986: copr: UTF8 functions should consider collation
Some [UTF8 functions](https://github.com/tikv/tikv/blob/master/components/tidb_query_vec_expr/src/impl_string.rs) in TiKV coprocessor do not properly take collation into consideration. For example, [`instr_utf8`](https://github.com/tikv/tikv/blob/44f189c39460ccec75c3f0a5d0210c82eeddc6fa/components/tidb_query_vec_expr/src/impl_string.rs#L561) now compares two strings by their lowercase representation, which is not the right definition.

Related functions:
- [ ] ord (may need to split into ord_utf8)
- [ ] ~~instr_utf8~~
- [x] find_in_set
- [x] strcmp
- [ ]  ~~Locate2ArgsUTF8~~
- [ ] ~~Locate3ArgsUTF8~~

### PR DESCRIPTION
**Pull request by TszKitLo40**
### What problem does this PR solve?

Issue Number: close #8986  

Problem Summary:

### What is changed and how it works?

What's Changed:
implement strcmp considering collation
### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 
- No release note

### CODE REVIEW

**Comment by Xuanwo** [comment]
Please sign-off your code to make DOC check happy.

File: components/tidb_query_expr/src/impl_string.rs [hunk 1]
```diff
@@ -679,9 +679,17 @@ pub fn substring_index(
 
 #[rpn_fn]
 #[inline]
-pub fn strcmp(left: BytesRef, right: BytesRef) -> Result<Option<i64>> {
+pub fn strcmp<C: Collator>(left: BytesRef, right: BytesRef) -> Result<Option<i64>> {
     use std::cmp::Ordering::*;
-    Ok(Some(match left.cmp(right) {
+    let left = match String::from_utf8(left.to_vec()) {
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_string.rs:684]: Convert to `String` requires extra copy, try convert to `str` like: `str::from_utf8(left)`

File: components/tidb_query_expr/src/impl_string.rs [hunk 2]
```diff
@@ -679,9 +679,17 @@ pub fn substring_index(
...
+        Ok(left) => left,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    let right = match String::from_utf8(right.to_vec()) {
+        Ok(right) => right,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    Ok(Some(match left.cmp(&right) {
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_string.rs:692]: We need to take `C::IS_CASE_INSENSITIVE` into consideration.

**Comment by TszKitLo40** [comment]
> Please sign-off your code to make DOC check happy.

How to trigger the DCO check again?

**Comment by Xuanwo** [comment]
Sign-off all commits requires a force-puch I guess.

File: components/tidb_query_expr/src/impl_string.rs [hunk 3]
```diff
@@ -679,9 +679,22 @@ pub fn substring_index(
...
-    Ok(Some(match left.cmp(right) {
+    let left = match str::from_utf8(left) {
+        Ok(left) => left,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    let right = match str::from_utf8(right) {
+        Ok(right) => right,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    let offset = if C::IS_CASE_INSENSITIVE {
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_string.rs:692]: `offset` is not a good name for result of `cmp`, please use `order` / `ord` instead.

File: components/tidb_query_expr/src/impl_string.rs [hunk 4]
```diff
@@ -3124,26 +3137,68 @@ mod tests {
     #[test]
     fn test_strcmp() {
         let test_cases = vec![
-            (Some(b"123".to_vec()), Some(b"123".to_vec()), Some(0)),
-            (Some(b"123".to_vec()), Some(b"1".to_vec()), Some(1)),
-            (Some(b"1".to_vec()), Some(b"123".to_vec()), Some(-1)),
-            (Some(b"123".to_vec()), Some(b"45".to_vec()), Some(-1)),
+            (
+                Some(b"123".to_vec()),
+                Some(b"123".to_vec()),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_string.rs:3142]: Please add a test case for `ABC`, `abc` with a CASE_INSENSITIVE collator.

Comment (TszKitLo40) [reply components/tidb_query_expr/src/impl_string.rs:3142]: > Sign-off all commits requires a force-puch I guess.

Done, PTAL

**Comment by TszKitLo40** [comment]
> Sign-off all commits requires a force-puch I guess.

Done, PTAL

File: components/tidb_query_expr/src/impl_string.rs (hunk 4 above)
Comment (TszKitLo40) [reply components/tidb_query_expr/src/impl_string.rs:3142]: Done

File: components/tidb_query_expr/src/impl_string.rs [hunk 5]
```diff
@@ -1118,7 +1131,7 @@ mod tests {
             (Some("TiKV"), Some(4i64)),
             (Some("あなたのことが好きです"), Some(33i64)),
             (Some("분산 데이터베이스"), Some(25i64)),
-            (Some("россия в мире  кубок"), Some(38i64)),
+            (Some("россия в мире  кубок"), Some(38i64)),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_string.rs:1134]: Please don't introduce unrelated changes.

Comment (TszKitLo40) [reply components/tidb_query_expr/src/impl_string.rs]: OK, I have fixed it.

**Review by Xuanwo — APPROVED** [review]

**Comment by Xuanwo** [comment]
ping @andylokandy , PTAL

File: components/tidb_query_expr/src/impl_string.rs [hunk 6]
```diff
@@ -679,9 +679,22 @@ pub fn substring_index(
...
+    let right = match str::from_utf8(right) {
+        Ok(right) => right,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    let order = if C::IS_CASE_INSENSITIVE {
+        left.to_lowercase().cmp(&right.to_lowercase())
+    } else {
+        left.cmp(&right)
+    };
+    Ok(Some(match order {
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_string.rs:697]:
```suggestion
    Ok(Some(match C::sort_compare(left, right) {
```

Comment (andylokandy) [reply components/tidb_query_expr/src/impl_string.rs:697]: I'm not sure if this complies.

Comment (TszKitLo40) [reply components/tidb_query_expr/src/impl_string.rs]: OK, I have used collator.

File: components/tidb_query_expr/src/impl_string.rs [hunk 7]
```diff
@@ -681,20 +681,7 @@ pub fn substring_index(
...
-        Ok(right) => right,
-        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
-    };
-    let order = if C::IS_CASE_INSENSITIVE {
-        left.to_lowercase().cmp(&right.to_lowercase())
-    } else {
-        left.cmp(&right)
-    };
-    Ok(Some(match order {
+    Ok(Some(match C::sort_compare(left, right).unwrap() {
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_string.rs:684]: Please don't `unwrap` here, give `copr` a chance to handle error.

File: components/tidb_query_expr/src/impl_string.rs [hunk 8]
```diff
@@ -679,9 +679,9 @@ pub fn substring_index(
 
 #[rpn_fn]
 #[inline]
-pub fn strcmp(left: BytesRef, right: BytesRef) -> Result<Option<i64>> {
+pub fn strcmp<C: Collator>(left: BytesRef, right: BytesRef) -> Result<Option<i64>> {
     use std::cmp::Ordering::*;
-    Ok(Some(match left.cmp(right) {
+    Ok(Some(match C::sort_compare(left, right).unwrap() {
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_string.rs:684]: Please use `?`. If the error type mismatched, then you need to `.map_err` before the `?`.

Comment (Xuanwo) [reply components/tidb_query_expr/src/impl_string.rs:684]: Maybe we can implement `From` for collection related error?

Comment (TszKitLo40) [reply components/tidb_query_expr/src/impl_string.rs]: After using `?`, will the error be handled?

**Review by Xuanwo — APPROVED** [review]

**Review by andylokandy — APPROVED** [review]
/merge

File: components/tidb_query_expr/src/impl_string.rs (hunk 8 above)
Comment (andylokandy) [reply components/tidb_query_expr/src/impl_string.rs]: Yes.

Comment (TszKitLo40) [reply components/tidb_query_expr/src/impl_string.rs]: It seems that the merge process has not been triggered.
